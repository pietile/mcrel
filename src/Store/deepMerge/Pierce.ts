// Unique pierce id
export const PierceID = Symbol('InsideID');

export interface Pierce<K, V> {
  id: typeof PierceID;
  key: K;
  value: V;
}

export function pierce<K, V>(key: K, value: V): Pierce<K, V> {
  return {
    id: PierceID,
    key,
    value,
  };
}
