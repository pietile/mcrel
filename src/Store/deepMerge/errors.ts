/* eslint-disable max-classes-per-file */

export class DeepMergeCircularError extends Error {
  constructor() {
    super('DeepMerge does not handle circular references.');
  }
}

export class PierceWrongTargetError extends Error {
  constructor(target: any) {
    super(`Target ${target} can't be pierced`);
  }
}

export class PierceWrongKeyError extends Error {
  constructor(key: any) {
    super(`Pierce error: Item with key ${key} not found`);
  }
}
