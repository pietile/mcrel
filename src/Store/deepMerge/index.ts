export * from './deepMerge';
export * from './errors';
export * from './Pierce';
export * from './Subset';
