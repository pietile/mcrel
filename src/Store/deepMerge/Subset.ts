import { Pierce } from './Pierce';

// Supported pierceable types
type Pierceable = ReadonlyArray<any> | ReadonlyMap<any, any>;

// Pierced object key type
type KeyType<T> = T extends Array<any> ? number : T extends Map<infer K, any> ? K : never;

// Pierced object value type
type ValueType<T> = T extends Array<infer R> ? R : T extends Map<any, infer V> ? V : never;

// Subset or Atomic
type SubsetObject<T, S, A> = T extends A ? Pick<T, keyof T> : Subset<T, S, A>;

// Check if S is a pierce
type SubsetPierceOrObject<T, S, A> = T extends Pierceable
  ? S extends Pierce<any, infer V>
    ? Pierce<KeyType<T>, Subset<ValueType<T>, V, A>>
    : SubsetObject<T, S, A>
  : SubsetObject<T, S, A>;

// Ensure S is a suptype of T with Atomics A
export type Subset<T, S, A> = {
  [K in keyof S]: K extends keyof T
    ? NonNullable<T[K]> extends object
      ? S[K] extends object // Go deep only if T[K] and S[K] are objects
        ? SubsetPierceOrObject<NonNullable<T[K]>, S[K], A>
        : T[K]
      : T[K]
    : never;
};
