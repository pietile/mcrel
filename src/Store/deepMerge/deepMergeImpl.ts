import { PierceWrongKeyError, PierceWrongTargetError } from './errors';
import { Pierce, PierceID } from './Pierce';

export type IsAtomicObject = (value: object) => boolean;

// DeepMerge implementation using runtime information only
export function deepMergeImpl(target: any, source: any, isAtomicObject: IsAtomicObject): any {
  if (typeof target !== 'object') {
    throw new Error(`merge target must be an object, but got: ${target}`);
  }

  if (typeof source !== 'object') {
    throw new Error(`merge source must be an object, but got: ${source}`);
  }

  const result = { ...target };

  Object.entries<any>(source).forEach(([key, value]): void => {
    if (typeof result[key] === 'object' && result[key] !== null && !!value) {
      if (typeof value === 'object' && value.id === PierceID) {
        result[key] = pierceMerge(target[key], value, isAtomicObject);

        return;
      }

      if (!isAtomicObject(result[key])) {
        result[key] = deepMergeImpl(target[key], value, isAtomicObject);

        return;
      }
    }

    result[key] = value;
  });

  return result;
}

// Piercing
function pierceMerge(target: any, pierce: Pierce<any, any>, isAtomicObject: IsAtomicObject): any {
  if (Array.isArray(target)) {
    const result = Array.from(target);

    if (result[pierce.key] === undefined) {
      throw new PierceWrongKeyError(pierce.key);
    }

    const { value } = deepMergeImpl(
      { value: result[pierce.key] },
      { value: pierce.value },
      isAtomicObject,
    );

    result[pierce.key] = value;

    return result;
  }

  if (target instanceof Map) {
    const result = new Map(target);

    if (!result.has(pierce.key)) {
      throw new PierceWrongKeyError(pierce.key);
    }

    const { value } = deepMergeImpl(
      { value: result.get(pierce.key) },
      { value: pierce.value },
      isAtomicObject,
    );

    result.set(pierce.key, value);

    return result;
  }

  throw new PierceWrongTargetError(target);
}
