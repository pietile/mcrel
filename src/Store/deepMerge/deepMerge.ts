import { GuardFunction, GuardedTypes } from '../GuardedTypes';

import { deepMergeImpl } from './deepMergeImpl';
import { DeepMergeCircularError } from './errors';
import { Subset } from './Subset';

// ES6 built-in Atomic types
type AtomicObject =
  | Date
  | Function
  | Promise<any>
  | ReadonlyArray<any>
  | ReadonlyMap<any, any>
  | ReadonlySet<any>
  | RegExp
  | WeakMap<any, any>
  | WeakSet<any>;

// Helper type that combines built-in Atomics and Atomics from guard functions
export type DeepMergeAtomics<G extends GuardFunction[]> = AtomicObject | GuardedTypes<G>;

// Merge source with target into new object
export function deepMerge<
  T,
  S extends Subset<T, S, DeepMergeAtomics<G>>,
  G extends GuardFunction[] = never
>(target: T, source: S, atomicGuards?: G): T {
  try {
    return deepMergeImpl(target, source, function isAtomicObject(value: object): boolean {
      const knwonAtomic =
        Array.isArray(value) ||
        value instanceof Date ||
        value instanceof Map ||
        value instanceof Promise ||
        value instanceof RegExp ||
        value instanceof Set ||
        value instanceof WeakMap ||
        value instanceof WeakSet;

      return knwonAtomic || (!!atomicGuards && atomicGuards.some((guard): boolean => guard(value)));
    });
  } catch (error) {
    if (
      (error.message && error.message.match(/stack|recursion/i)) ||
      error.number === -2146828260
    ) {
      /**
       * Browsers give this different errors name and messages:
       * chrome/safari: "RangeError", "Maximum call stack size exceeded"
       * firefox: "InternalError", too much recursion"
       * edge: "Error", "Out of stack space"
       */

      throw new DeepMergeCircularError();
    }

    throw error;
  }
}
