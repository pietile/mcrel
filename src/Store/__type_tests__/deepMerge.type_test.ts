import { deepMerge, pierce } from '../deepMerge';

interface Atomic {
  type: 'unique';
  a: number;
  subObject: {
    b: string;
  };
}

interface SubState {
  a: number;
  subObj: {
    b: number;
    c?: string;
  };
}

interface State {
  a: number;
  b: string;
  subObj1: {
    c: number;
  };
  subObj2: {
    d: number;
  } | null;
  subObj3?: {
    d: number;
  };
  array: number[];
  map: Map<number, string>;
  set: Set<number | string>;
  atomic: Atomic;
  array2: SubState[];
  map2: Map<string, SubState>;
}

const obj: State = {
  a: 0,
  b: 'b',
  subObj1: {
    c: 0,
  },
  subObj2: {
    d: 0,
  },
  array: [1],
  map: new Map(),
  set: new Set([1, 'a']),
  atomic: {
    type: 'unique',
    a: 0,
    subObject: {
      b: 'b',
    },
  },
  array2: [],
  map2: new Map(),
};

function isAtomic(value: any): value is Atomic {
  return value.type === 'unique';
}

// @dts-jest:fail:snap Unknown field
deepMerge(obj, { c: 1 });

// @dts-jest:fail:snap Field type
deepMerge(obj, { b: 1 });

// @dts-jest:fail:snap null to NonNullable
deepMerge(obj, { subObj1: null });

// @dts-jest:fail:snap Incorrect subObj
deepMerge(obj, { subObj2: { a: 1 } });

// @dts-jest:fail:snap undefined to null
deepMerge(obj, { subObj2: undefined });

// @dts-jest:fail:snap value to nullable obj
deepMerge(obj, { subObj2: '' });

// @dts-jest:fail:snap null to undefined
deepMerge(obj, { subObj3: null });

// @dts-jest:fail:snap value to optional obj
deepMerge(obj, { subObj3: 1 });

// @dts-jest:fail:snap Map value type
deepMerge(obj, { array: new Map([[1, 1]]) });

// @dts-jest:fail:snap Array value type
deepMerge(obj, { array: [''] });

// @dts-jest:fail:snap Map key type
deepMerge(obj, { map: new Map([['1', '1']]) });

// @dts-jest:fail:snap Set value type
deepMerge(obj, { set: new Set([false]) });

deepMerge(
  obj,
  {
    // @dts-jest:fail:snap Custom Atomic value type
    atomic: {
      type: 'unique',
      subObject: {
        b: 'b',
      },
    },
  },
  [isAtomic],
);

// @dts-jest:fail:snap Pierce wrong target
deepMerge(obj, { atomic: pierce('a', 1) });

// @dts-jest:fail:snap Pierce array wrong key type
deepMerge(obj, { array2: pierce('a', { a: 1 }) });

// @dts-jest:fail:snap Pierce array wrong value type
deepMerge(obj, { array2: pierce(1, { b: null }) });

// @dts-jest:fail:snap Pierce map wrong key type
deepMerge(obj, { map2: pierce(1, { subObj: { b: 1 } }) });

// @dts-jest:fail:snap Pierce map wrong value type
deepMerge(obj, { map2: pierce('a', { subObj: { c: 1 } }) });
