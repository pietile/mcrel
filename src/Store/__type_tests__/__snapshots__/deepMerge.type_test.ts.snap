// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Array value type (type) should match snapshot 1`] = `"Type 'string' is not assignable to type 'number'."`;

exports[`Custom Atomic value type (type) should match snapshot 1`] = `"Property 'a' is missing in type '{ type: string; subObject: { b: string; }; }' but required in type 'Pick<Atomic, \\"a\\" | \\"type\\" | \\"subObject\\">'."`;

exports[`Field type (type) should match snapshot 1`] = `"Type 'number' is not assignable to type 'string'."`;

exports[`Incorrect subObj (type) should match snapshot 1`] = `"Type 'number' is not assignable to type 'never'."`;

exports[`Map key type (type) should match snapshot 1`] = `
"Type 'Map<string, string>' is not assignable to type 'Pick<Map<number, string>, \\"set\\" | \\"forEach\\" | \\"entries\\" | \\"keys\\" | \\"values\\" | \\"clear\\" | \\"delete\\" | \\"get\\" | \\"has\\" | \\"size\\">'.
  Types of property 'set' are incompatible.
    Type '(key: string, value: string) => Map<string, string>' is not assignable to type '(key: number, value: string) => Map<number, string>'.
      Types of parameters 'key' and 'key' are incompatible.
        Type 'number' is not assignable to type 'string'."
`;

exports[`Map value type (type) should match snapshot 1`] = `"Type 'Map<number, number>' is missing the following properties from type 'Pick<number[], number | \\"map\\" | \\"length\\" | \\"toString\\" | \\"toLocaleString\\" | \\"pop\\" | \\"push\\" | \\"concat\\" | \\"join\\" | \\"reverse\\" | \\"shift\\" | \\"slice\\" | \\"sort\\" | \\"splice\\" | \\"unshift\\" | \\"indexOf\\" | ... 14 more ... | \\"includes\\">': map, length, pop, push, and 20 more."`;

exports[`Pierce array wrong key type (type) should match snapshot 1`] = `
"Type 'Pierce<string, { a: number; }>' is not assignable to type 'Pierce<number, Subset<SubState, { a: number; }, AtomicObject>>'.
  Type 'string' is not assignable to type 'number'."
`;

exports[`Pierce array wrong value type (type) should match snapshot 1`] = `
"Type 'Pierce<number, { b: null; }>' is not assignable to type 'Pierce<number, Subset<SubState, { b: null; }, AtomicObject>>'.
  Type '{ b: null; }' is not assignable to type 'Subset<SubState, { b: null; }, AtomicObject>'.
    Types of property 'b' are incompatible.
      Type 'null' is not assignable to type 'never'."
`;

exports[`Pierce map wrong key type (type) should match snapshot 1`] = `
"Type 'Pierce<number, { subObj: { b: number; }; }>' is not assignable to type 'Pierce<string, Subset<SubState, { subObj: { b: number; }; }, AtomicObject>>'.
  Type 'number' is not assignable to type 'string'."
`;

exports[`Pierce map wrong value type (type) should match snapshot 1`] = `
"Type 'Pierce<string, { subObj: { c: number; }; }>' is not assignable to type 'Pierce<string, Subset<SubState, { subObj: { c: number; }; }, AtomicObject>>'.
  Type '{ subObj: { c: number; }; }' is not assignable to type 'Subset<SubState, { subObj: { c: number; }; }, AtomicObject>'.
    The types of 'subObj.c' are incompatible between these types.
      Type 'number' is not assignable to type 'string | undefined'."
`;

exports[`Pierce wrong target (type) should match snapshot 1`] = `
"Type 'Pierce<string, number>' is not assignable to type 'Subset<Atomic, Pierce<string, number>, AtomicObject>'.
  Types of property 'id' are incompatible.
    Type 'unique symbol' is not assignable to type 'never'."
`;

exports[`Set value type (type) should match snapshot 1`] = `
"Type 'Set<boolean>' is not assignable to type 'Pick<Set<string | number>, \\"forEach\\" | \\"entries\\" | \\"keys\\" | \\"values\\" | \\"clear\\" | \\"delete\\" | \\"has\\" | \\"size\\" | \\"add\\">'.
  Types of property 'forEach' are incompatible.
    Type '(callbackfn: (value: boolean, value2: boolean, set: Set<boolean>) => void, thisArg?: any) => void' is not assignable to type '(callbackfn: (value: string | number, value2: string | number, set: Set<string | number>) => void, thisArg?: any) => void'.
      Types of parameters 'callbackfn' and 'callbackfn' are incompatible.
        Types of parameters 'value' and 'value' are incompatible.
          Type 'boolean' is not assignable to type 'string | number'."
`;

exports[`Unknown field (type) should match snapshot 1`] = `"Type 'number' is not assignable to type 'never'."`;

exports[`null to NonNullable (type) should match snapshot 1`] = `"Type 'null' is not assignable to type '{ c: number; }'."`;

exports[`null to undefined (type) should match snapshot 1`] = `"Type 'null' is not assignable to type '{ d: number; } | undefined'."`;

exports[`undefined to null (type) should match snapshot 1`] = `"Type 'undefined' is not assignable to type '{ d: number; } | null'."`;

exports[`value to nullable obj (type) should match snapshot 1`] = `"Type 'string' is not assignable to type '{ d: number; } | null'."`;

exports[`value to optional obj (type) should match snapshot 1`] = `"Type 'number' is not assignable to type '{ d: number; } | undefined'."`;
